### ISSUE TITLE
Bug: [no-unnecessary-condition] false positive with `??=` operator and `exactOptionalPropertyTypes` compiler option

### ISSUE THREAD

**Issue by chharvey** [typescript-eslint/typescript-eslint#6635]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.5&sourceType=module&code=MYewdgzgLgBCBGArAXDA3jADgJxJg-KgNpgCuAtvAKbYC6MAvjALzoMDcAUJwGaljAoAS3AwIVKAHkBVABQA3AIYAbUlVRlKNAJTFN1Ouk4wY2CaWxg4SAHQ48MfPlZElqqrXYwA9N5gBCGAABKABPTCoIYGwhTCgAWkjlITAobzAQeP4wKmBIiEVsUPjQMAATIWFwVAAiAFUwHLyIAqKYUoqqsBUAGhgqAA8IwSoymGUqHgSAC0VysSEyqjgeGAADJzW4COxFKBBsGH2YaiwQFqF4ZVCYMmVlOEP+JZ4U0ZsazgZOIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK+SpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoAkNZOgB4CG0euBxTqACprIljwBPACpCBZSnkwBXdABpwEKCWnJWAORmpUAYQAW6aAGtJYaXKUBfEFaA

### Repro Code

```typescript
const obj: { prop?: [number] } = {};

function setOnce(value: number): [number] {
    return obj.prop ??= [value]; // ! @typescript-eslint/no-unnecessary-condition: "Unnecessary conditional, expected left-hand side of `??` operator to be possibly null or undefined."
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "exactOptionalPropertyTypes": true,
    "strictNullChecks": true
  }
}
```

### Expected Result

I expected the expression `obj.prop ??= [value]` to not report an error.

Since the property hasn’t been set yet, it is actually undefined at runtime.

### Actual Result

The expression `obj.prop ??= [value]` is reported as an error.

typescript-eslint thinks `obj.prop` is non-nullish since `exactOptionalPropertyTypes` is on. We don’t want to turn it off because we still don’t want to allow statements like `obj.prop = undefined;`.

### Additional Info

Related: #3553, #5344, https://github.com/typescript-eslint/typescript-eslint/issues/6632

**Comment by bradzacher** [comment]
Filed a bug with TS - https://github.com/microsoft/TypeScript/issues/53305
IMO this should be fixed in TS itself cos even TS behaves differently with assignment operators vs their expanded form. But we'll see what they say.

If not, then we either must implement our own logic to determine if the LHS might be an exact optional type or not, or we might just disable checks on `??=`. First port of call is the types, then we can make a decision about next steps.

**Comment by chharvey** [comment]
@bradzacher — I believe TS is working as intended since after assignment it will be `number` and not `undefined`. We’ll see what they say. Regardless, I would add this function to your playground, for completeness:
```ts
function ee() {
xx.prop = 1;
//  ^?
// typeof xx.prop === number ❌
}
```

**Comment by bradzacher** [comment]
@chharvey if it was working as intended then `+=` and `??=` would work the same way.

`(number | undefined) + number === number` always (because `NaN` is a `number`).
So by your reasoning:
```ts
xx.prop += 1
//  ^?
// typeof xx.prop === number, because the result of the assignment operator will always be number
```

But that's not what we see from TS - we see that TS correctly shows that `typeof xx.prop === number | undefined`.
Which is why it's inconsistent!

You need to separate the read from the write!
`xx.prop` has not been updated until after the assignment node is complete and traversed.
i.e.
```
xx.prop += 1;
            ^ semi = end of expression = assignment complete
              before this point xx.prop has not been updated
              so any intellisense hovers should be the type *before* the write
```

but [from the looks of it](https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true#code/DYUwLgBAHgXBB2BXAtgIxAJwgHwgZzAwEt4BzHCReAExADMSRqBuAKFYk6lYHoeIAegH52naBAC8EAIxs+gkRy5z+w0V0kQA5FpUL10XqqFA) TS does some funny stuff with the type of a variable that has a union type - it appears to invalidate the refinements on the variable at specific write times.

**Comment by rgant** [comment]
This is filling up my code with a bunch of `// eslint-disable`s. Anything that can be done about this?

**Comment by laverdet** [comment]
This issue hits our team really bad. We're currently working around this with a local patch via `pnpm`:

```diff
diff --git a/patches/@typescript-eslint__eslint-plugin@5.58.0.patch b/patches/@typescript-eslint__eslint-plugin@5.58.0.patch
new file mode 100644
index 000000000..c38427dcd
--- /dev/null
+++ b/patches/@typescript-eslint__eslint-plugin@5.58.0.patch
@@ -0,0 +1,12 @@
+diff --git a/dist/rules/no-unnecessary-condition.js b/dist/rules/no-unnecessary-condition.js
+index 6d36b811889c22e0b6dc61beb40a82671fd47ad4..d0f06ace52ecbffe4b571ebc9d05450dca15a1fb 100644
+--- a/dist/rules/no-unnecessary-condition.js
++++ b/dist/rules/no-unnecessary-condition.js
+@@ -479,7 +479,6 @@ exports.default = (0, util_1.createRule)({
+             }
+         }
+         return {
+-            AssignmentExpression: checkAssignmentExpression,
+             BinaryExpression: checkIfBinaryExpressionIsNecessaryConditional,
+             CallExpression: checkCallExpression,
+             ConditionalExpression: (node) => checkNode(node.test),
```

**Comment by karlhorky** [comment]
It seems like the TypeScript PR https://github.com/microsoft/TypeScript/pull/54777 was merged, which closed the issue in the TypeScript repo:

- https://github.com/microsoft/TypeScript/issues/53305#event-10100428639

Will this mean that this issue will "fix itself" once the TS PR changes are released in a version of TypeScript (eg. maybe 5.2)?

**Comment by QuentinLemCode** [comment]
Hello
One of my co-worker had the same issue with this piece of code : 
https://typescript-eslint.io/play/#ts=5.2.2&showAST=ts&fileType=.tsx&code=C4TwDgpgBAYghgYwsKBeKBvAUFKBrASwDsATAfgC4oBnYAJ2IHMsBfLLUSKAcToHsArmDSYcUAGaJk1KvCTAA2gF1W7BHyK0oAWxC9BYGT35DlIhWIwSpwIwquFSVAOTA%2Bb51BYAaTCyVe3pbW8nYOxCQuwATRnj5%2BAWwqWOqaKOJ8fCK6%2BkLUYgB04gA2cMAAsnBgABTVMRDaAJRoAHxQ9dpFNtSNhXQQJAJItf3UAsXAvggCdP1EwABqcMUCEM2obdi4uATiUNXTsxDzSysQBY4kzdhi21AA9PdQABYQ-VCCdFDqJNAZdEhqDpoG4oHASCQoMBXlAiHwALQCIhECCA6hwOggeGpEgxAgaW7baqjcaKQ5zRbLVYXCIBMhkdDKRoFMACajPA4zCmnVa9O5sQn9YAzIhQEkTMTxDAsMFAgBKqL4dBIAB5aAwiIxfHJkMoWo0gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSLQ%2Bg9ABMAkkVXoAHmoCC0aOhLCwAMwCGqEuikBfEPaA&tokens=false

Can you confirm it's related ?

**Comment by stianjensen** [comment]
> Will this mean that this issue will "fix itself" once the TS PR changes are released in a version of TypeScript (eg. maybe 5.2)?

Running with 5.3-beta now seems to still trigger the warning for me, so I'm assuming some fix is also needed here?
